### PR TITLE
Class-wise saliency map generation for the detection task

### PR DESCRIPTION
**Pull request by negvet**
Enable class-wise saliency map generation for the detection case.

Tested with all detection models from `external\model-preparation-algorithm\configs\detection\` on PASCAL VOC dataset

related PR:
https://github.com/openvinotoolkit/model_preparation_algorithm/pull/97

Some examples (mobilenetv2_ssd):
image

### CODE REVIEW

**Comment by negvet** [comment]
run ote-test

File: external/mmdetection/detection_tasks/apis/detection/openvino_task.py [hunk 1]
```diff
@@ -576,18 +577,43 @@ def infer(
...
+                        roi=dataset_item.roi
+                    )
+                    dataset_item.append_metadata_item(saliency_media, model=self.model)
+                elif saliency_map.ndim == 3:
+                    # Multiple saliency maps per image (class-wise saliency map)
+                    labels = self.task_environment.get_labels(include_empty=True)
+                    num_saliency_maps = saliency_map.shape[0]
+                    if num_saliency_maps == len(labels) + 1:
+                        # include the background as the last label
+                        labels.append(LabelEntity('backgroung', Domain.DETECTION))
```
Comment (sungmanc) [inline external/mmdetection/detection_tasks/apis/detection/openvino_task.py]: Maybe want to `background` ?

Comment (negvet) [reply external/mmdetection/detection_tasks/apis/detection/openvino_task.py]: Thanks. Fixed!

**Comment by negvet** [comment]
run ote-test

**Review by wonjuleee — DISMISSED** [review]
LGTM

File: external/mmdetection/detection_tasks/apis/detection/openvino_task.py [hunk 2]
```diff
@@ -576,18 +577,43 @@ def infer(
...
+                    if num_saliency_maps == len(labels) + 1:
+                        # Include the background as the last label
+                        labels.append(LabelEntity('background', Domain.DETECTION))
+                    for class_id, class_wise_saliency_map in enumerate(saliency_map):
+                        actmap = get_actmap(
+                            class_wise_saliency_map, (dataset_item.width, dataset_item.height)
+                        )
+                        class_name_str = labels[class_id].name
+                        saliency_media = ResultMediaEntity(
+                            name=class_name_str,
```
Comment (goodsong81) [inline external/mmdetection/detection_tasks/apis/detection/openvino_task.py]: Why don't you align with classification?
image

**Review by goodsong81 — CHANGES_REQUESTED** [review]
Thank you for the PR.
The test you included is a kind of unit test. Would you create separate PRs for integration tests which will cover Geti1.1 validation?
It might be just to verify # saliency map via OTE OV task API call.
cc: @chuneuny-emily

**Comment by goodsong81** [comment]
Ah, will you create additional PRs for limiting # of saliency map? (top N classes, etc)

File: external/mmdetection/detection_tasks/apis/detection/openvino_task.py (hunk 2 above)
Comment (negvet) [reply external/mmdetection/detection_tasks/apis/detection/openvino_task.py]: Thanks for a notice. 
Geti UI already have static `Saliency map` phrase, so it is clear for the user that there are saliency maps.
Thus, it was decided to replace `f"Saliency Map: {name=class_name_str}"` with `name=class_name_str`.
So that in the drop down menu we have just the names of the the classes.

I will align classification with detection.

**Comment by negvet** [comment]
> Ah, will you create additional PRs for limiting # of saliency map? (top N classes, etc)

IMO, OTE should support the most advanced use case, when saliency maps are provided for all classes. And the customer can simplify it by selecting and displaying the most important saliency maps. [this is at least for the current `task.infer` interface, when saliency maps are generated while regular inference]
What do you think? 

I discuss the limiting # of saliency map to be done on the Geti side.

**Comment by negvet** [comment]
run ote-test

**Comment by negvet** [comment]
> Thank you for the PR. The test you included is a kind of unit test. Would you create separate PRs for integration tests which will cover Geti1.1 validation? It might be just to verify # saliency map via OTE OV task API call. cc: @chuneuny-emily

Thank you for the comment. Sure, I will create such tests.
Not sure that it will be within this PR. Is it ok?

**Review by goodsong81 — DISMISSED** [review]
LGTM. Thanks!

**Comment by goodsong81** [comment]
> > Thank you for the PR. The test you included is a kind of unit test. Would you create separate PRs for integration tests which will cover Geti1.1 validation? It might be just to verify # saliency map via OTE OV task API call. cc: @chuneuny-emily
> 
> Thank you for the comment. Sure, I will create such tests. Not sure that it will be within this PR. Is it ok?

Sure! :)

**Review by sungmanc — DISMISSED** [review]
LGTM. 
Although I think the XAI algorithm didn't have dependencies with different model architectures, It could be better if unit tests cover all of our templates.
(In this case, only ResNet-18 and MobileNetv2_w1 SSD are used for test cases for classification and detection, respectively)

**Comment by negvet** [comment]
> LGTM. Although I think the XAI algorithm didn't have dependencies with different model architectures, It could be better if unit tests cover all of our templates. (In this case, only ResNet-18 and MobileNetv2_w1 SSD are used for test cases for classification and detection, respectively)

Good suggestion. I agree, will add it. (together with Geti1.1 validation tests in a separate PR probably)

File: external/mmdetection/detection_tasks/apis/detection/openvino_task.py [hunk 3]
```diff
@@ -456,17 +457,48 @@ def infer(
...
+                    labels = self.task_environment.get_labels()
+                    num_saliency_maps = saliency_map.shape[0]
+                    if num_saliency_maps == len(labels) + 1:
+                        # Include the background as the last category
+                        labels.append(LabelEntity('background', Domain.DETECTION))
+                    for class_id, class_wise_saliency_map in enumerate(saliency_map):
+                        class_name_str = labels[class_id].name
+                        if class_name_str in predicted_class_set:
+                            # TODO (negvet): Support more advanced use case,
+                            #  when all/configurable set of saliency maps is returned
```
Comment (negvet) [inline external/mmdetection/detection_tasks/apis/detection/openvino_task.py:488]: I was requested to make only predicted classes to be explained for Geti (at least for Geti1.1).
This is due to possible memory overhead issues in the many classes use case.
I apply it only for OV tasks.

Also update cls in this PR (I know, I better do it in a separate PR, sorry - I am in a bit hurry)

@goodsong81 @sungmanc @wonjuleee please take a look

**Comment by negvet** [comment]
run ote-test

**Comment by dongkwan-kim01** [comment]
LGTM. The class selection part can be modified in the future but it seems okay with the current version.
We can improve it by only calculating the selected class's saliency map in the MPA side

- Test in Geti with latest commit, YOLOX, trained 30 images with [dog, cat, horse] class
image

File: external/mmdetection/detection_tasks/apis/detection/openvino_task.py (hunk 3 above)
Comment (wonjuleee) [reply external/mmdetection/detection_tasks/apis/detection/openvino_task.py:488]: Okay, it looks like reasonable.

**Review by wonjuleee — APPROVED** [review]

File: external/mmdetection/detection_tasks/apis/detection/openvino_task.py (hunk 3 above)
Comment (sungmanc) [reply external/mmdetection/detection_tasks/apis/detection/openvino_task.py:488]: LGTM

**Review by sungmanc — APPROVED** [review]

**Comment by harimkang** [comment]
Hi @negvet , This seems to be a problem with the current instance-segmentation. Could you please check it out? (all ins-seg test FAILED)
https://ci-ote.iotg.sclab.intel.com/job/ote/job/pr-ote-test/804/
image


https://github.com/openvinotoolkit/model_preparation_algorithm/blob/a99aea0296a6c1987c014d55c26dfb56ca8be611/mpa/modules/models/detectors/sam_detector_mixin.py#L24
